juspay/hyperswitch
# feat(router): add outgoing payment webhooks for v2

## Type of Change



## Description

This PR adds outgoing payment webhooks for v2. This PR improves the webhook delivery flow by introducing traits to enhance extensibility and modularity. Additionally, it optimizes the raise_webhooks_analytics_event process by removing an unnecessary database fetch.

### Additional Changes





## Motivation and Context



## How did you test it?

Tested Manually
Setup incoming webhooks and trigger a successful payment, to receive an outgoing webhook

 - The incoming webhook to be setup to {base_url}/v2/webhooks/(merchant_id}/{profile_id}/{mca_id}
 - Create Payment 
```
curl --location 'http://localhost:8080/v2/payments/create-intent' \
--header 'api-key: dev_gkzPh8C1IaaBlf1Vo31QNoKUkhRuKUDYfm6zoe1LLiZL3rpOyDppUVI2nrLT9KIP' \
--header 'Content-Type: application/json' \
--header 'x-profile-id: pro_7MjgLAahgfPro5dOWlHk' \
--header 'Authorization: api-key=dev_gkzPh8C1IaaBlf1Vo31QNoKUkhRuKUDYfm6zoe1LLiZL3rpOyDppUVI2nrLT9KIP' \
--data-raw '{
    "amount_details": {
        "order_amount": 100,
        "currency": "USD"
    },
    
    "capture_method":"automatic",
    "authentication_type": "no_three_ds",
    "billing": {
        "address": {
            "first_name": "John",
            "last_name": "Dough"
        },
        "email": "example@example.com"
    },
    "shipping": {
        "address": {
            "first_name": "John",
            "last_name": "Dough",
            "city": "Karwar",
            "zip": "581301",
            "state": "Karnataka"
        },
        "email": "example@example.com"
    }
}'
```
- Confirm Payment
```
curl --location 'http://localhost:8080/v2/payments/12345_pay_0196803f8a8e7840810bf8dcc6b0fa74/confirm-intent' \
--header 'x-client-secret: cs_0196803f8aa376c282d881ceb57c6aac' \
--header 'x-profile-id: pro_7MjgLAahgfPro5dOWlHk' \
--header 'Content-Type: application/json' \
--header 'Authorization: publishable-key=pk_dev_2c342eea3f5f404bbb6938fc00bae5e7, client-secret=cs_0196803f8aa376c282d881ceb57c6aac' \
--data '{
    "payment_method_data": {
        "card": {
            "card_number": "4242424242424242",
            "card_exp_month": "10",
            "card_exp_year": "25",
            "card_holder_name": "John Doe",
            "card_cvc": "100"
        }
    },
    "payment_method_type": "card",
    "payment_method_subtype": "credit"
}
'
```
Outgoing Payload Response
```
{
  "merchant_id": "cloth_seller_pKhmzgXrMESoKLtxaSSd",
  "event_id": "evt_0196803f9be67fb28967b94a04807324",
  "event_type": "payment_succeeded",
  "content": {
    "type": "payment_details",
    "object": {
      "id": "12345_pay_0196803f8a8e7840810bf8dcc6b0fa74",
      "status": "succeeded",
      "amount": {
        "order_amount": 100,
        "currency": "USD",
        "shipping_cost": null,
        "order_tax_amount": null,
        "skip_external_tax_calculation": "Skip",
        "skip_surcharge_calculation": "Skip",
        "surcharge_amount": null,
        "tax_on_surcharge": null,
        "net_amount": 100,
        "amount_to_capture": null,
        "amount_capturable": 0,
        "amount_captured": null
      },
      "customer_id":NULL
      "connector":"stripe"
      "created":"2025-04-29T06:32:50.077Z"
      "payment_method_data":{...}
      "payment_method_type":"card"
      "payment_method_subtype":"credit"
      "connector_transaction_id":"pi_3RJ7iqD5R7gDAGff0jchh4In"
      "connector_reference_id":NULL
      "merchant_connector_id":"mca_5Ekmol3FgPzd98j0c3wY"
      "browser_info":NULL
      "error":NULL
      "shipping":{...}
      "billing":{...}
      "attempts":NULL
      "connector_token_details":{...}
      "payment_method_id":NULL
      "next_action":NULL
      "return_url":"https://google.com/success"
      "authentication_type":"no_three_ds"
      "authentication_type_applied":NULL
      }
    }
  },
  "timestamp": "2024-11-19T12:42:10.267Z"
}
```


Also, tested failed webhook delivery cases
## Checklist
